sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightFoldingBuilder.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.squareup.sqldelight.intellij.lang
+
+import com.alecstrong.sqlite.psi.core.psi.SqliteTypes
+import com.intellij.lang.ASTNode
+import com.intellij.lang.folding.FoldingBuilder
+import com.intellij.lang.folding.FoldingDescriptor
+import com.intellij.lang.folding.NamedFoldingDescriptor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.project.DumbAware
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiWhiteSpace
+import com.squareup.sqldelight.core.SqlDelightTypes
+import com.squareup.sqldelight.core.psi.SqlDelightStmtIdentifier
+import com.squareup.sqldelight.intellij.util.childOfType
+import com.squareup.sqldelight.intellij.util.childrenWithType
+import com.squareup.sqldelight.intellij.util.nextLeafOrNull
+import com.squareup.sqldelight.intellij.util.prevSiblingOfType
+
+class SqlDelightFoldingBuilder : FoldingBuilder, DumbAware {
+
+  override fun buildFoldRegions(root: ASTNode, document: Document): Array<FoldingDescriptor> {
+    return root.createFoldingDescriptors()
+  }
+
+  private fun ASTNode.createFoldingDescriptors(): Array<FoldingDescriptor> {
+    val descriptors = mutableListOf<FoldingDescriptor>()
+    val importElements = mutableListOf<PsiElement>()
+    val statementLists = childrenWithType(SqlDelightTypes.SQL_STMT_LIST)
+    for (statementList in statementLists) {
+      for (statement in statementList.getChildren(null)) {
```

Comment:
`statementLists.flatMap { it.getChildren(null) }.forEach { statement ->` to de-nest a little bit